### PR TITLE
feat: improve pg write queue

### PR DESCRIPTION
**Pull request by bestmike007**
## Description

This PR is to fix the potential issue described in https://medium.com/@alkor_shikyaro/transactions-and-promises-in-node-js-ca5a3aeb6b74

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Does this introduce a breaking change?

No.

## Are documentation updates required?

No. 

## Testing information

Current tests covered this.

## Checklist
- [x] Code is commented where needed
- [x] Unit test coverage for new or modified code paths
- [x] `npm run test` passes
- [ ] Changelog is updated
- [x] Tag 1 of @rafaelcr or @zone117x for review

### CODE REVIEW

**Comment by bestmike007** [comment]
@zone117x

File: src/datastore/helpers.ts [hunk 1]
```diff
@@ -1343,14 +1343,26 @@ export function newReOrgUpdatedEntities(): ReOrgUpdatedEntities {
...
   enqueue(task: Parameters<PQueue['add']>[0]): void {
-    void this.queue.add(task);
+    const p = this.queue.add(task);
+    p.catch(void 0);
+    this.tasks.push(p);
   }
-  done(): Promise<void> {
-    return this.queue.onIdle();
+  async done(): Promise<void> {
+    // https://medium.com/@alkor_shikyaro/transactions-and-promises-in-node-js-ca5a3aeb6b74
```
Comment (rafaelcr) [inline src/datastore/helpers.ts:1358]: I don't believe the problem described in this post applies to the tasks handled by this queue for a couple of reasons:
1. `PQueue`'s `onIdle` [already behaves like `allSettled`](https://www.npmjs.com/package/p-queue#onidle)
    > Returns a promise that settles when the queue becomes empty, and all promises have completed
1. The API only enqueues work to this queue inside of a pre-existing SQL transaction (all the block processing code is contained in a single `sqlWriteTransaction`). If there's ever an exception thrown inside, the transaction aborts and rolls back all the pending writes. We've done extensive tests on this and it's worked correctly for a long time.

**Comment by rafaelcr** [comment]
@bestmike007 I'm curious to now if this fix was submitted because of a problem you're experiencing? We'd be happy to test some edge case if you've found one we haven't covered yet

File: src/datastore/helpers.ts (hunk 1 above)
Comment (bestmike007) [reply src/datastore/helpers.ts:1358]: 1. That's what I thought, but it was wrong; here's a demonstration: https://replit.com/@bestmike007/TroubledEarnestSandboxes#index.ts
2. `onIdle` does not throw an exception, it simply waits all promises to complete.

What I'm seeing is, there was an unhandledRejection and then the process restarted.

**Comment by bestmike007** [comment]
Yes I was trying to resolve a problem, our nodes run into unrecoverable state, not sure if it's related to this PR.

Here's one of the situations I saw, which should not be happening:

telegram-cloud-photo-size-5-6273870458726235191-y

Another one, I couldn't found a screenshot, was reporting something like `table tid from new index tuple (13374,63) overlaps with invalid duplicate tuple at offset 4 of block 6211 in inde`, and can be fixed by 3 steps: remove the `nft_custody_unique` constraints, remove duplicates with lower block heights, and then rebuild the unique constraint.

File: src/datastore/helpers.ts [hunk 2]
```diff
@@ -1343,14 +1343,26 @@ export function newReOrgUpdatedEntities(): ReOrgUpdatedEntities {
...
+  private tasks: Promise<unknown>[];
   constructor() {
     const concurrency = Math.max(1, getUintEnvOrDefault('STACKS_BLOCK_DATA_INSERT_CONCURRENCY', 4));
     this.queue = new PQueue({ concurrency, autoStart: true });
+    this.tasks = [];
   }
   enqueue(task: Parameters<PQueue['add']>[0]): void {
-    void this.queue.add(task);
+    const p = this.queue.add(task);
+    p.catch(e => logger.error(e, 'PgWriteQueue task failed'));
```
[the remark this answers is not in the document]
Comment (bestmike007) [reply src/datastore/helpers.ts:1354]: If there's no catch, the process will be terminated by an unhandledRejection.

The error is logged here because `throw firstRejected.reason` only throws the first one, we don't have any log for the other errors.

[the remark this answers is not in the document]
Comment (zone117x) [reply src/datastore/helpers.ts:1354]: Ah okay, would it make sense to use the p-queue error event instead? https://www.npmjs.com/package/p-queue#error

**Comment by bestmike007** [comment]
> Hmm I don't believe that error is related to this change. It seems strange, though, because the ON CONFLICT ON CONSTRAINT is there precisely to avoid this error. Can you tell me more about your setup? Is this error occurring in an event replay, or a normal Stacks node sync?

It it not related, it's another issue I saw when the node keeps crashing because of unhandledRejection when it fails to process a new_block event.

Our nodes are set up using the script in this repo: https://github.com/alexgo-io/stacks-node-mainnet, and nodes restoring from both hiro archive and the cold backup saw this issue.

File: src/datastore/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bestmike007) [reply src/datastore/helpers.ts:1354]: This is what we saw using a node built from this PR.

image

[the remark this answers is not in the document]
Comment (zone117x) [reply src/datastore/helpers.ts:1354]: I see, we can manually capture the errors ourself with `.catch` to prevent a `terminated by an unhandledRejection`. But could se instead use the build-in p-queue errror handler, e.g. 
```
queue.on('error', error => {
	console.error(error);
});
```
?

[the remark this answers is not in the document]
Comment (bestmike007) [reply src/datastore/helpers.ts:1354]: This is not available before v7.0.0: https://github.com/sindresorhus/p-queue/tree/v6.6.2?tab=readme-ov-file#events
